src/Middleware.php

```diff
@@ -186,7 +186,24 @@ public static function retry(
             return new RetryMiddleware($decider, $delay, $handler, $stats);
         };
     }
-
+    /**
+    * Middleware wrapper function that adds an invocation id header to requests.
+    * This is a uniquely generated UUID to identify initial and subsequent retries as part of a complete request lifecycle.
+    *
+    * @return callable
+    */
+    public static function invocationID()
+    {
+        return function (callable $handler) {
+            return function (
+                CommandInterface $command,
+                RequestInterface $request = null
```

Comment:
This middleware will throw an error if `$request` is null. You could make `$request` non-nullable and mention that it must be added after the build step in the documentation block, as we do in [`Aws\S3\ApplyChecksumMiddleware`](https://github.com/aws/aws-sdk-php/blob/master/src/S3/ApplyChecksumMiddleware.php#L9).